### ISSUE TITLE
[Linux] Long match loading times resulting in abandon streak

### ISSUE THREAD

**Issue by tukkek** [ValveSoftware/Dota-2#1337]
I have never ever before gotten into low priority in Dota but in the last couple days since the Dueling Fates update I have been put into low priority twice in a row due to abandons. I must have had around 5-10 abandons at the very most in 700 hours of playing Dota but in the last couple days since the update I must have had another 5 in a very short time period. This is how bad the problem is.

I understand if anyone comes to me and say "well failing to load is bad for other players so you should be put into low priority for bad conduct for that" - but the point is I'm not having a bad conduct, it's because **the software takes so long to load that it causes a disconnect and then an abandon**. The first match I tried to join today I had to reconnect not once but twice - because the time to reconnect was so long that the server kicked me out before I could actually reconnect. 

I will post my system requirements next and from my account here of barely having any meaningful amount of disconnects prior to this particular week, I hope you'll understand that the problem is not with my hardware. I also don't have any programs running in background and being a programmer I can tell it shouldn't be anything related to my system in particular - especially since I had never had any trouble in this regard before.

Lastly, I would like to point out that I don't think this is a problem with the Dueling Fates update in particular. I have noticed that whenever an update comes out, the first match I load (always "all pick" for me, except for these LP games) seems to take a lot longer than usual. This is probably the new assets from each patch being processed during the pick phase? The thing is that with the Dueling Fates update and the large number of daily patches since then this has probably grown a previous issue (bad loading times on Linux) from a mere annoyance to an actual severe gameplay issue (putting people on LP for no fault of their own).

I will try to pre-load a singleplayer lobby match in the future whenever I start the Dota client to see if this helps but at this point I'm just working-around the real issue here. Please consider that **being put in Low Priority is a very severe penalty** that should be reserved for actual consistent below-acceptable display of bad conduct. It's very hard getting out from LP, especially if you get put on it a few times in a row: you can't play your favorite modes or heroes while at it and a single win can take you 4 or 5 games due to disconnects, bad teammates, etc - literally hours of your life for a single win to get you out of LP. It feels extremely aggravating and makes me want to quit Dota even as a paying customer. Please consider this report as an important issue (for me, if not for anyone else, though I imagine this must affect a lot of other people too) and let me know what steps I can take to help you diagnose and fix this problem.

Just to make it clear: I don't expect the LP system to be fixed but something must be done about these loading times in Linux. I'm consistently getting very long slow-downs during picking phase as the game pre-loads, to the point I'll eventually receive the "communication with the server has been disrupted" message, which forces me to try to reconnect, which itself leads to a whole new batch of loading and possible repeat disconnect after that. **This seems to affect in particular the first match played after opening the Dota client and in special after the game has just been updated.** Other matches played in the same session don't suffer from this problem (or at least not enough as to cause more disconnects and abandons).

Thank you again for the excellent support I've received in this bug tracker and I hope this can get resolved soon.

**Comment by tukkek** [comment]
My Steam system information. I'm pretty sure I'm well-covered for playing Dota but feel free to peruse it and make your assessment. My graphics card in particular is very low-tier but it plays the game fine on modest settings - I don't think it should be impacting pre-loading times during the picking phase at all, in itself? Other than that I'm pretty sure I have an average to above-average setup here.

```
Computer Information:
    Manufacturer:  Unknown
    Model:  Unknown
    Form Factor: Desktop
    No Touch Input Detected

Processor Information:
    CPU Vendor:  AuthenticAMD
    CPU Brand:  AMD FX-8320E Eight-Core Processor              
    CPU Family:  0x15
    CPU Model:  0x2
    CPU Stepping:  0x0
    CPU Type:  0x0
    Speed:  3200 Mhz
    8 logical processors
    8 physical processors
    HyperThreading:  Unsupported
    FCMOV:  Supported
    SSE2:  Supported
    SSE3:  Supported
    SSSE3:  Supported
    SSE4a:  Supported
    SSE41:  Supported
    SSE42:  Supported
    AES:  Supported
    AVX:  Supported
    CMPXCHG16B:  Supported
    LAHF/SAHF:  Supported
    PrefetchW:  Unsupported

Operating System Version:
    Debian GNU/Linux testing (buster) (64 bit)
    Kernel Name:  Linux
    Kernel Version:  4.13.0-1-amd64
    X Server Vendor:  The X.Org Foundation
    X Server Release:  11903000
    X Window Manager:  KWin
    Steam Runtime Version:  steam-runtime-beta-release_2017-07-24

Video Card:
    Driver:  NVIDIA Corporation GeForce GT 710/PCIe/SSE2
    Driver Version:  4.5.0 NVIDIA 375.82
    OpenGL Version: 4.5
    Desktop Color Depth: 24 bits per pixel
    Monitor Refresh Rate: 59 Hz
    VendorID:  0x10de
    DeviceID:  0x128b
    Revision Not Detected
    Number of Monitors:  1
    Number of Logical Video Cards:  1
    Primary Display Resolution:  1366 x 768
    Desktop Resolution: 1366 x 768
    Primary Display Size: 16.14" x 9.06" (18.50" diag)
                                            41.0cm x 23.0cm (47.0cm diag)
    Primary Bus: PCI Express 8x
    Primary VRAM: 2048 MB
    Supported MSAA Modes:  2x 4x 8x 16x 

Sound card:
    Audio device: Realtek ALC887-VD

Memory:
    RAM:  3937 Mb

Miscellaneous:
    UI Language:  English
    LANG:  en_US.UTF-8
    Total Hard Disk Space Available:  253416 Mb
    Largest Free Hard Disk Block:  66290 Mb
    VR Headset: None detected

Recent Failure Reports:
```

**Comment by gdrewb-valve** [comment]
Are you using OpenGL or Vulkan?  What are your launch options?

**Comment by tukkek** [comment]
Hello again, Drew - and thank you for all the previous fixes! My launch options are simply `-console`, which I have now removed since I only added it for curiosity's sake a while ago. My renderer is whatever the default for the game is - I tried opening the client's Video settings tab to see for sure which one it is but there doesn't seem to be any information there about it. Let me know how I can find out for sure.

Just to add more information: since writing this report and getting out of LP twice I have started opening a lobby to make sure (at least some) assets are preloaded before jumping into matchmaking. This has partial results. One time at least it still disconnected me after the picking phase (*when loading the map* and not *during the picking phase itself* as before) and at least one other time my first matchmaking experience was smooth after following this "use case" (start Dota client -> immediately create and start lobby -> play All Pick matchmaking) - or at least as much as this workaround constitutes a legitimate "use case" :D anyway it seems to have improved the issue a little bit since I am not currently in LP as I write this but it's also far from perfect. I haven't tested this extensively but these partial results are interesting enough to share, hopefully.

**Comment by zpostfacto** [comment]
Seems like there are potentially a bunch of problems here.  The grace times to load into the map are supposed to be exceedingly generous.  Even on a low end system it really shouldn't take very load to load into hero selection.  Then it continues to load the rest of the map during hero selection, and then if everybody selects their hero before your client finishes loading, there is another period where the server waits for you to finish.  Both of those timeouts are supposed to be really generous.

Also I don't understand why you are getting "disconnected".  You should be able to stay connected to the server while loading.  There is no reason why a long load time would cause a network disconnection.  (Unless the server is intentionally kicking you because it gave up waiting.)  But that would only happen once.  After that, I wouldn't expect you to be able to reconnect and play (without penalty at least).  So either we have a bug, or the server is kicking you and it just isn't clear that this is what is happening, or something else.

We just really need a specific example with a complete console log on your computer, from the time you booted your client.  It should have a MatchID in it which will allow us to locate the corresponding log on the server.

**Comment by tukkek** [comment]
Hello Fletcher and thank you for your time and input taken into analyzing this issue. I have tried to make it as clear as possible at which points the slowdowns occur but from you comment I understand how I failed to do so properly, I will try to rectify it by being more explicit:

1. I have never had a problem getting *into* the hero selection screen. This always loads fine every time.
2. I understand that the latest Dota client pre-loads the game assets while heroes are being picked. This seems to be the first issue - as heroes are being picked I notice my "reading from hard drive" light going pretty much full-on and the game begins to slow down. This can and will often result in me getting the "communication with the server has been disrupted" message. I'll expand on it in a little bit.
3. After all heroes are picked (again, I play All Pick exclusively, except for the recent LP matches which are Single Draft) then the "show match" screen appears (for lack of better term, when all 10 heroes are shown and some wild drums start playing). I assume that from here on out the match is still finishing loading. Disconnects can happen in-between this phase - and also even after a very long freeze, more rarely, the match will appear fully loaded for a split second and then disconnect me back to the client's home screen (probably a "finished loading, let's chat with the server, oh well looks like it has kicked us" scenario).

I believe the issue happens much more on step 3 (after the drums stop playing and the game is still loading), a large majority of the time - but it can happen in 2 as well, and that's also when the slowdown becomes very noticeable (at least for the first game I play after starting the client). I am talking minutes of the black "finishing loading..." screen before the in-game interface will show.

Take what I will say with a pound of salt because I have no sense of any of the game's internals, but as a programmer maybe my mental image will help you understand (even though it's much more likely wrong). It's almost as if with each hero picked during the picking phase, a new thread is spawned to process the specific hero's assets - except that at some point there are too many threads competing for memory and CPU time that the system just slows to a halt. This is what if "feels" like, having worked with threads and processes a lot myself - I hope it helps you understand the issue, even if that's not really what happens. I have a 8 core CPU too so maybe that's a somewhat unconventional edge case that isn't properly tested.

I have no idea why I'm being disconnected and allowed back into the game if you're saying it shouldn't happen (but boy am I glad that it allows me to, otherwise I wouldn't be able to play, probably). My best guess is that the aforementioned "halt" from too much processing and I/O "soft-locks" whatever socket is handling networking and thus the server just thinks I timed out and would allow me to reconnect (as I believe typical timeouts due to network congestion or temporary Internet failure would always allow a player to reconnect).

Finally, for the part that you asked me to provide a full log.. well, I'd love to Fletcher, but I can't do it because it means another abandon and who knows how many hours in low priority. You'll understand I'm not looking forward to it yet again, after I've been put there twice this week without actually engaging in bad behavior as far as intentionally abandoning games is concerned. The low priority experience is extremely aggravating for me personally and I will not willingly put myself into it for the sake of a bug report - I definitely would if it didn't require so many games and hours to get out of, not to mention long-term damage to my in-game behavior score, which can result in even more LP time. I'm really sorry, let me know if there is some way we can work around it if you guys can pardon me "behind the screens" for the abandon or whatever.

I would just like to repeat again that even though loading has always been clunky I have never had as many abandons as I did last week or any problems other than some slow performance while loading my first match, as I explained already. As you said yourself, even a low-specs computer should be able to handle the startup fine, so there's definitely something weird going on, I think. Also, I forgot to mention before: I take extra care not to have any other typical processes running in parallel to the game so as to guarantee it has as much memory or CPU time as possible so it can try to load as smoothly as possible in order to prevent an abandon.

Sorry for the long texts I'm just trying to be thorough  so that this has a higher chance of being successfully understood and resolved. Thanks again and let me know what else I can do to help! Thank you both!

**Comment by zpostfacto** [comment]
If you make an effort to reproduce this issue and get those logs, I'll fix any negative impact it has on your account!  This is a really big problem, I'd like to fix it.

Make sure you have a reliable way to capture the full console log before you begin.

**Comment by tukkek** [comment]
Thank you Fletcher, I'm happy to hear that this is as relevant an issue to you as it is for me. I'll probably leave it for Monday though so I can contact you if/when I get an abandon or put into LP. Can you instruct me which file has the console output that I should send over? I'm pretty sure there's one that's rewritten per-session?

Meanwhile I found out I can reproduce the slowdown and general system unresponsiveness by loading a local lobby match will bots (which I would assume tries loading each hero's assets as well). Loading a lobby all by myself is a lot smoother, in comparison.

The Dota client is doing massive I/O reads (and negligible writes) all the while from the start of the picking screen to the end of the "finishing loading..." screen - and even after that, but then I imagine that's just the 9 default "hard" bots playing the game. I'm talking non-stop reads from start to finish while letting every single timer run down on its own and even after that in the "finishing loading..." screen. I have no idea what the client is doing but that's some massive amount of reading and processing in each and all of my 8 CPUs, non-stop, plus I/O reads non-stop as well...

I have also assessed that my memory is never full so it shouldn't be an issue of reading/writing from/to my Linux swap memory partition which would be my first guess. There also isn't enough I/O writes that would be accompanied by all that memory swapping.

Anyway all this extra info is probably meaningless without the console output but anyway it's here already in any case if you need to look at it after I get you a log file.

Oh I also thought maybe this could be the Steam overlay running amok for some reason but the only process I have that is using any significant amount of CPU meanwhile is Dota. Unfortunately it must use threads for this processing and not entirely new processes so my system monitor can't give me any extra info on what it is actually going on inside the client.

I have timed the entire process from beginning of picking phase to the moment where the in-game UI appears at first (while letting picking timers run out) for my All Pick match with bots and that's pretty much exactly 3 minutes of high-intensity processing and non-stop I/O reads.

Again, I'm not sure what Dota is doing but whatever it is, it's doing a lot of it! This is what my CPU usage looks like throughout:

![spectacle d22844](https://user-images.githubusercontent.com/3819826/32391644-ea0dada6-c0b9-11e7-8f85-fc0cab0d4d76.png)

Anyway thanks for the patience and I'll try to come up with a log for this issue next week.

**Comment by gdrewb-valve** [comment]
If you put -con_logfile in your launch options dota will create a console.log file in your dota install directory under the game/dota subdir.

**Comment by zpostfacto** [comment]
It doesn't seem unusual for a bunch of disk IO to be going on while loading the map.  But I do find it quite unexpected that loading a map is able to keep EIGHT CPUs so busy.  I would expect most of the time to simply be spent waiting on the disk.

**Comment by zpostfacto** [comment]
You might try just using "perf top" while loading into a lobby filled with bots, to try to figure out what it's doing.

http://man7.org/linux/man-pages/man1/perf-top.1.html

**Comment by tukkek** [comment]
Alright I figured out how to recreate this issue safely (without using matchmaking). Steps:

1. Open Dota client, create a lobby, fill it with bots, host it remotely (Brazil region for me).
2. Enter lobby until the hero picking screen becomes responsive (but don't do anything with it).
3. Simultaneously launch a bunch of heavy programs (in my case: `audacity & gimp & ktorrent & ~/bin/eclipse/eclipse`). Here this results in the system becoming unresponsive from anywhere to a couple to several minutes (which I assume is mostly because of Dota hogging all CPUs). I also have a couple of program running in the background (Firefox and System Monitor) - but the general idea is to overload your system to reproduce the issue with Dota.
4. As each of these external programs open up, also close them to keep the CPU busy.
5. After all the external programs have booted up and were then closed by you, go back to the Dota window and you'll find the error message mentioned in my first report:

![screenshot_20171103_174615](https://user-images.githubusercontent.com/3819826/32393125-803ec99a-c0bf-11e7-9625-86ffee09541b.png)

I know this is literally forcing the client to fail to load properly but it seems to do the trick. I didn't try reconnecting (apparently the game wouldn't allow me to for solo bot lobbies?) - but hopefully with this reproducible test case it should be easier to look further into the whole thing. Let me know if there anything else I should do to help in investigating this issue. 

Here's the error log: 

[condump000.txt](https://github.com/ValveSoftware/Dota-2/files/1442406/condump000.txt)

**Comment by tukkek** [comment]
> It doesn't seem unusual for a bunch of disk IO to be going on while loading the map

Agreed but if you think of 3 full minutes of 6MB/s reading then you have 1GB of data loaded. That's pretty much the entire size of the Dota process, which sounds entirely excessive.

Of course, take that with a grain of salt since I'm not sure what Linux considers I/O reads (it probably even puts network use there?) and I'm just taking the number I saw flashing across the screen in my tests instead of actually measuring anything properly.. anyway, it still seems very excessive...

OK now that I got you a error log I should step away and let the big guns work this out :D I have definitely overstayed my welcome here, if by word count alone and not anything else haha thank you again, hope this helps improve the Linux client and let me know if I can help in any other way!

**Comment by zpostfacto** [comment]
OK, looking at that log, the networking was definitely getting pushed around by all of the activity, and not getting timely service.  You can see it complaining about it.  However, it didn't actually timeout at the transport level.  There was a disconnection at the source engine layer because communication at the *application* layer did not happen.  That is a bug.  This has been a problem for a while.  Thanks for sharing this info!

**Comment by kisak-valve** [comment]
With those symptoms, keep an eye on swap usage, a tool like dstat would give you near-realtime transfer rates in and out of swap.

**Comment by tukkek** [comment]
As I mentioned before my system still had at least half a gig of free memory while I was performing initial tests (without extra stress) with my system monitor. I also thought swap might be an issue but apparently it isn't. Probably more related to too many threads, their management by the OS and all the extra overhead it implies (context switches galore). I posted a screen cap today showing the CPU usage is all over the place.

Another sign this isn't a swap issue is that although the I/O read count was massive, the write count was negligible... but I don't blame you for missing this information too, previously in the thread. Anyway, Fletcher seems to have found a lead for this - just let me know if I should run more tests locally if it turns out to not be whatever is causing the main problem here (the abandons themselves).

**Comment by tukkek** [comment]
This is my CPU usage with Dota after it failed to load a solo lobby without bots (the simplest game mode I can think of) after disconnecting with the "communication disrupted" message, which I've mentioned before. I'll say it again to drive the point home: I am currently sitting at the home screen, disconnected from a match and Dota is furiously processing in parallel on 8 CPUs. It has been running like this for many minutes now.

![dota idle](https://user-images.githubusercontent.com/3819826/32416576-9951c382-c232-11e7-8d27-a0f52c3b907c.png)

Again, I have no other processes running on the CPUs, have free memory on my system, etc, etc.

@fletcherdvalve I think it has become pretty apparent that whatever is the underlying issue here, it is not the application-layer timeout. The only way I can explain this is that threads are spawned during the pre-loading phase (picking phase) and they are either stuck on an endless loop or just failing to terminate. My best guess is that this is causing the application-layer timeout - which as you explain, seems to be an important issue of its own, but not the real reason why my system crawls to a halt while loading the map.

EDIT: I'm 99% sure I was running a remote lobby when this happened so I don't think all the processing could be a local server running the game in the background, even if I myself was disconnected from it as a payer.

EDIT 2: I'm having a little trouble coming up with a test case that reproduces this scenario so it might just end up being circumstantial evidence and not very reliable or dependable, unfortunately.

**Comment by tukkek** [comment]
I really hate to put it like this but Dota has become very close to unplayable for me. Even though I'm loading a solo match prior to matchmaking, every time I do get into matchmaking I get disconnected and manage only to connect again close to 4:00 on the game clock - not only this alone is awful for a competitive game, meaning I start what could be easily be a hour-long game at a huge disadvantage (and thus put my entire team at a disadvantage as well), it's also extremely close to getting a new abandon. Loading times were bad before the Dueling Fates update but now every single game I try to join is flirting with the danger of getting back into low-priority queue (yes, for the third time this week, even if I have never ever had been put there before).

I hope I have raised enough red flags and provided you with enough evidence that there is something massively wrong with the Dota client on Linux. If I get into low priority again (which, frankly, right now, sounds like only a matter of time), I will just give up on Dota and wait for the incoming free-to-play Starcraft to run well enough on Wine as my new e-sports fix. I'm sorry but I can't keep doing Dota like this, even as a paying customer and recent battle pass owner.

Here is my latest log. Loaded a solo match first to get as much loading as possible out of the way, then tried joining a ranked match, got kicked out (after the map screen was fully loaded), took close to five minutes to reconnecting again (staring at the "finishing loading..." screen) but somehow didn't get an abandon even though I must have been as close as a few seconds away from it, as I understand it.

[condump001.txt](https://github.com/ValveSoftware/Dota-2/files/1444285/condump001.txt)

I apologize if I sound dramatic but like many others I was excited for this patch but instead it lowered my game experience literally straight down into forced, unwarranted low-priority queue. I understand this is no one's fault in particular and that maintaining a port can be a hard task in itself but in the end the only thing that matters for me as a player is that the game I was enjoying well enough a couple weeks ago suddenly isn't living up to any standard of quality or genuine entertainment. I wish you all the best of luck in identifying and fixing this issue and hope that you'll consider it a top priority as I cannot imagine I'm the only Linux user suffering from it, especially when I got a decent enough computer and I can't imagine being able to play Dota at all were it a step or two below in processing power, given these severe loading issues.

**Comment by gdrewb-valve** [comment]
If you get a chance can you do a solo lobby and try "perf top" as fletcherdvalve mentioned above when the game is sitting there burning CPU?  You won't have symbols so you won't get useful stacks but it might show a little info as to what code the threads are running.

**Comment by tukkek** [comment]
I am currently in Low Priority queue again. Steam derped for a moment (disconnected with "no Steam logon" message in the middle of a game, which happened to me a couple of times only in my 700 hours playing Dota) and the loading time for reconnect was way too long - I got into the match again just in time to see the "has stayed disconnected for too long and will be assessed with an abandon" broadcast notification.

Not even a week old, Dueling Fates has managed to put me into low priority 3 times for no fault of my own. Again, I'm really sorry to say it but at this point in time it's just matter-of-fact to state that the Linux port is literally unplayable for 7.07 and I'm done with it until this issue is fully resolved.

It might not mean anything but I'm changing my Steam review to negative as well for the time being.

**Comment by tukkek** [comment]
@gdrewb-valve `perf top` shows me some render stuff and nvidia-driver related processing when I open Dota but other than that nothing I can see that would be relevant, either on the loading screen or while the game is running.
